packages/react-reconciler/src/ReactFiberThrow.js

```diff
@@ -175,6 +177,150 @@ function attachPingListener(
   }
 }
 
+export function unstable_avoidThisRenderImpl(
+  root: FiberRoot,
+  returnFiber: Fiber,
+  sourceFiber: Fiber,
+  value: Thenable,
+  renderExpirationTime: ExpirationTime,
+) {
+  // should we just return outside non-blocking mode?
+  if ((sourceFiber.mode & BlockingMode) === NoMode) {
+    return;
+  }
+
+  // this function body was copy pasted from throwException(), and then changed
+
+  // mark the thenable as "special"
+  // let's do it this ugh way right now, use a weakset or something later
+  value.__isAvoidedThenable = true;
+
+  const thenable: Thenable = (value: any);
+
+  checkForWrongSuspensePriorityInDEV(sourceFiber);
+
+  let hasInvisibleParentBoundary = hasSuspenseContext(
+    suspenseStackCursor.current,
+    (InvisibleParentSuspenseContext: SuspenseContext),
+  );
+
+  // Schedule the nearest Suspense to re-render the timed out view.
+  let workInProgress = returnFiber;
+  do {
+    if (
+      workInProgress.tag === SuspenseComponent &&
+      shouldCaptureSuspense(workInProgress, hasInvisibleParentBoundary)
+    ) {
+      // Found the nearest boundary.
+
+      // Stash the promise on the boundary fiber. If the boundary times out, we'll
+      // attach another listener to flip the boundary back to its normal state.
+
+      const thenables: Set<Thenable> = (workInProgress.updateQueue: any);
+      if (thenables === null) {
+        const updateQueue = (new Set(): any);
+        updateQueue.add(thenable);
+        workInProgress.updateQueue = updateQueue;
+      } else {
+        thenables.add(thenable);
+      }
+
+      if (hasInvisibleParentBoundary && (workInProgress.effectTag & SoftCapture === NoEffect)) {
+        // this implies "first phase"
+        attachPingListener(root, renderExpirationTime, thenable);
+        // what happens if an image resolves before the tree is committed?
+        // it should 'invalidate' the tree (or not) and effectively restart
+        workInProgress.effectTag |= SoftCapture;
+
+        // how is this different from markUpdateTimeFromFiberToRoot
+        workInProgress.expirationTime = renderExpirationTime;
+        // set didAvoidRenderTime?
+        if(!workInProgress.memoizedState){
+          workInProgress.memoizedState = {
+            hydrated: null, 
+            retryTime: renderExpirationTime, // ???
+            didAvoidRenderTime: renderExpirationTime // ???
+          }
+        }
+        // in commit, DON'T add retry listeners
+      } else {
+        // should this be uncommented?
+        // workInProgress.effectTag |= SoftCapture;
+        
+        // in commit, add retry listener
+      }
+
+      return;
+    }
+    // This boundary already captured during this render. Continue to the next
+    // boundary.
+    workInProgress = workInProgress.return;
+  } while (workInProgress !== null);
+  // No boundary was found. Fallthrough to error mode.
```

Comment:
This is an error case. Very rare, not worth optimizing. I would just throw here.

Reply:
Added a warning instead.